src/transformers/models/gptsan_japanese/modeling_gptsan_japanese.py

```diff
@@ -0,0 +1,1326 @@
+            https://github.com/tanreinama/GPTSAN/blob/main/report/model.md
+
+        Returns:
+            `MoEModelOutputWithPastAndCrossAttentions` or `tuple` if `return_dict` returns
+            MoEModelOutputWithPastAndCrossAttentions insted of tuple
+        """
+        return_dict = return_dict if return_dict is not None else self.config.return_dict
+        device = self.position_embeddings.weight.device
+        if input_ids is None:
+            input_ids = torch.zeros([1, 1]).int().to(device)
+        num_pasts_contexts = 0
+        num_batch = input_ids.shape[0]
+        pasts_or_squad_value = None
+        if past_key_values is not None:
+            num_pasts_contexts = past_key_values[0][0].shape[2]
+        elif self.config.d_spout and spout is not None:
+            num_pasts_contexts = 1
+
+        if num_precontext is not None:
+            if not (
+                len(num_precontext.shape) == 2 and num_precontext.shape[1] == 1
+            ):  # num_precontext Should be [batch,1]
+                raise ValueError("num_precontext should be [batch, 1] size.")
+            num_precontext = torch.reshape(num_precontext, [-1])
+        else:
+            num_precontext = torch.zeros([num_batch]).int().to(device)
+
+        num_input_contexts = input_ids.shape[1]
+        num_output_contexts = num_input_contexts + num_pasts_contexts
+
+        hidden_states = self.embed_tokens(input_ids)
+
+        if past_key_values is not None:
+            pasts_or_squad_value = past_key_values
+        elif self.config.d_spout and spout is not None:
+            pasts_or_squad_value = self.spout(spout)
+            pasts_or_squad_value = torch.reshape(
+                pasts_or_squad_value,
+                [
+                    num_batch,
+                    self.config.num_layers,
+                    2,
+                    self.config.num_heads,
+                    num_pasts_contexts,
+                    self.config.d_model // self.config.num_heads,
+                ],
+            )
+            pasts_or_squad_value = torch.split(pasts_or_squad_value, [1] * self.config.num_layers, dim=1)
+            pasts_or_squad_value = tuple(
+                torch.split(a.squeeze(1), [1, 1], dim=1).squeeze(1) for a in pasts_or_squad_value
+            )
+        else:
+            pasts_or_squad_value = [None] * self.config.num_layers
+
+        token_position = torch.arange(num_input_contexts).to(device) + num_pasts_contexts
+        token_position = torch.clip(token_position, num_pasts_contexts, self.config.max_position_embeddings - 1)
+
+        gather_position = (
+            (torch.zeros((self.config.d_model, num_input_contexts)).to(device) + token_position).transpose(0, 1).long()
+        )
+        hidden_states += torch.gather(self.position_embeddings.weight, dim=0, index=gather_position)
+
+        attention_mask = (
+            make_variable_prefix_area_mask(num_input_contexts, num_output_contexts, num_precontext) * -10000.0
```

Comment:
```suggestion
            make_variable_prefix_area_mask(num_input_contexts, num_output_contexts, num_precontext) * torch.finfo(hidden_states.dtype).min
```
I think using the dtype prevents potential issues with 8bits and float16, correct me if I am wrong.